Approved/implicitly-scoped-qubit-allocation.md

```diff
@@ -0,0 +1,342 @@
+
+```qsharp
+operation DifferentLifetime() : Result {
+    mutable r = Zero;
+    {
+        use q = Qubit();
+        // Apply operations...
+        set r = M(q);
+        // q is released here.
+    }
+    // Classical post-processing...
+    return r;
+}
+```
+
+# Implementation
+
+The proposal could be implemented in one of two ways:
+
+1. As syntactic sugar for the existing block statements.
+   The new `use` and `borrow` statements are transformed in the AST into `using` and `borrowing` block statements, taking in the remaining statements of the block automatically.
+   The scope statement could be transformed into `if (true) { }`.
+   This would require no change to code generation.
+2. As new AST nodes, similar to `let` and `mutable` nodes for classical values.
+   A new scope statement AST node is added.
+   Code generation must release qubits at the end of the scope, but this is already analogous to reference counting for certain data types like arrays in QIR.
+
+Both options seem similar in terms of complexity and implementation timeline, but option 2 seems like the cleaner solution.
+Neither option has any performance differences compared to the existing implementation.
+
+## Timeline
+
+There are no dependencies on other proposals or libraries.
+The work needed for either option appears to be minimal, since it is similar to the existing implementation.
+
+# Further Considerations
+
+## Related Mechanisms
+
+This proposal relies on the existing qubit management functionality that is already a core feature of Q#.
+It does not change the behavior of qubit management; it only provides new syntax for existing functionality.
+
+## Impact on Existing Mechanisms
+
+The scope statement added in this proposal can also be used to limit the scope of non-qubit variable bindings declared with `let` and `mutable`.
+This may be useful for limiting the lifetime of large data structures like arrays or for otherwise communicating the intent that a variable should only be used in a specific section within a callable.
+
+## Anticipated Interactions with Future Modifications
+
+Because the proposed modification is similar to existing functionality, it is not expected to have any impact on future language development.
+
+## Alternatives
+
+### Alternative 1: Add new statements while preserving existing block syntax
+
+Q# could add non-block `using` and `borrowing` statements while preserving existing syntax:
+
+```qsharp
+using q = Qubit();
+// ...
+
+using (q = Qubit()) {
+    // ...
+}
+
+borrowing q = Qubit();
+// ...
+
+borrowing (q = Qubit()) {
+    // ...
+}
+```
+
+### Alternative 2: Add compound using and borrowing block statements
+
+Instead of adding non-block statements, Q# could extend the existing block statement syntax to allow multiple `using` and `borrowing` statements in a row that have only one block attached:
+
+```qsharp
+using (a = Qubit())
+borrowing (b = Qubit())
+using (c = Qubit()) {
+    // ...
+}
+```
+
```

Comment:
If I recall, there was a third alternative discussed at one point, where the `use` keyword would appear on the right-hand side:
```Q#
let register = use Qubit[2 * n + 1];
```
It may be helpful to compare to that alternative as well, especially as I think the proposed syntax presents some real advantages over putting `use` on the right.

Reply:
I hadn't seen that proposal before, but as I read through this I could help but think of this approach as well. It doesn't seem like there is really much of a distinction between `use q = Qubit()` and `let q = Qubit()`. The only thing you really would want the `use` for is to distinguish between whether or not you are getting a "clean" qubit or not, i.e. whether or not you are borrowing the qubit. To that end the `use` and `borrow` sort of take the place of the `new` keyword from other languages: `let q = use Qubit();` `let w = borrow Qubit();`.

Reply:
And actually at that point we may as well just use the `new` keyword instead of `use`, as it is not only familiar to developers but it makes more sense in this context purely from reading the statement out loud. The `borrow` keyword here would still be useful and makes sense: `let q = new Qubit();` and `let w = borrow Qubit();`.

Reply:
My issue with putting `use` on the right is that we don't want to allow this:

```qsharp
use a = use Qubit();
mutable b = use Qubit();
set b = a; // Is the original qubit of b released here?
```

or this:

```qsharp
Foo(use Qubit()); // When is the qubit released? The scope is unclear.
```

That means we need more complicated rules:

1. Qubit initializers can only appear in a `use` expression.
2. `use` expressions can only appear on the right-hand side of a `let` statement.

Compare to: qubit initializers can only appear on the right-hand side of a `use` statement.

About `new Qubit()`, I think that Q# tries to avoid the connotation that qubits are OOP-style objects. Operations on an allocated qubit can implicitly affect the entire quantum state, so they aren't new, independent objects like the `new` keyword would imply.

Reply:
I'm probably missing some context here, but why wouldn't they always be completely independent at the moment of the assignment statement? Talking specifically about `new`/`use` and not `borrow`.

Reply:
We could simply disallow mutable variables holding qubits. Is there something about using a `let` that means we also have to allow `mutable` for the same situation? If so, that would really be the only difference between a `let` and a left-side `use`: the connotation that the `let` could be replaced for a `mutable`. But I don't expect that that connotation is especially true or that we couldn't simple disallow it in this case.

Reply:
> I'm probably missing some context here, but why wouldn't they always be completely independent at the moment of the assignment statement?

Phase kickback, for example:

```qsharp
@EntryPoint()
operation Main() : Unit {
    using (a = Qubit()) {
        FlipPhase();
        DumpMachine();
    }
}

operation FlipPhase() : Unit {
    using (b = Qubit()) {
        within {
            X(b);
        } apply {
            Z(b);
        }
    }
}
```

`FlipPhase()` changes `a` from |0> to -|0> by only applying operations to `b`.

Reply:
About `use` on the right, I think my argument wasn't quite right - if `use Qubit()` is one unit, then the rule is that it can only appear on the right of a `let` statement, similar to how `Qubit()` can only appear on the right of a `use` statement, so that is not too bad. The other issue I have though, is that the lifetime of a qubit is tied to the scope of a variable - `use q = Qubit()` makes it clear that when `q` goes out of scope, the qubit is released. `let q = use Qubit()` is not as clear what the lifetime is IMO - `q` looks like a normal, non-resource variable.

For comparison, F# also does `use x = foo ()`, not `let x = use foo ()`, and C# does `using var x = Foo();`, not `var x = using Foo();`.

Reply:
Is the only difference then that you can't have a return statement with a variable declared with `use`?

Reply:
I guess I'm trying to see the fundamental difference in logic between the `use` and the `let` statements.

Reply:
The difference is that the value of variables declared with `use` become invalid outside the scope, while the value of variables declared with `let` don't.

```qsharp
mutable numbers = EmptyArray<Int>();
{
    let xs = [1, 2, 3];
    set numbers = xs;
}
Message($"{numbers[2]}");
```

works OK because while `xs` is out of scope, the array it pointed to is still valid. But

```qsharp
mutable qubits = EmptyArray<Qubit>();
{
    use qs = Qubit[3];
    set qubits = qs;
}
X(qubits[2]);
Message($"{MResetZ(qubits[2])}");
```

doesn't work; even though we tried to save the qubit array, the qubits in the array are no longer valid because `qs` went out of scope.

Reply:
It's pretty much the same as F# or C# -- something you "create" with `use` is `IDisposable` and gets disposed of (the `Dispose` method gets called) when it goes out of scope. We don't have interfaces, but qubits are the Q# equivalent of disposable.
In all 3 languages, the `use` keyword on the left indicates that there's special processing of the *value* (not the variable!) related to when the variable goes out of scope.

Reply:
So then the `use` is really more attached to the value of the qubit(s) rather than the container, doesn't it make more sense then to have it closer to the value as in `let q = use Qubit();` rather than closer to the container in `use q = Qubit();`? Also the differentiating scenario you posted is prevented when we prevent mutables from holding qubit references, as previously suggested. To that end, we could look at a scenario where qubit references are returned from a callable:
```
operation Foo() : Qubit {
    using (q = Qubit()) { return q; }
    // or
    use q = Qubit();
    return q;
    // or
    let q = use Qubit();
    return q;
}
```
I believe this scenario is currently allowed, which to me begs the question of what is supposed to happens when this operation is called, but this could also be prevented by not allowing the Qubit type to be anywhere in the return type of a callable. Of course the other approach would be to be more permissive with the Qubit type and allow its scoping rules and lifetime to be the same as other types.

Reply:
> So then the `use` is really more attached to the value of the qubit(s) rather than the container

Well, it's both. When the container goes out of scope, the value it contains invalidated. That's why I think it's more important to emphasize the container than the value, because the container decides what happens to the value and when.

> I believe this scenario is currently allowed, which to me begs the question of what is supposed to happens when this operation is called

It's undefined. The qubit is not valid after it's returned. Yes, ideally the compiler would prevent this.

> this could also be prevented by not allowing the Qubit type to be anywhere in the return type of a callable

That is too strict. This is a well-defined function:

```qsharp
function FlipPair(q1 : Qubit, q2 : Qubit) : (Qubit, Qubit) {
    return (q2, q1);
}
```

> the other approach would be to be more permissive with the Qubit type and allow its scoping rules and lifetime to be the same as other types

Qubit lifetimes being scope-limited is a deliberate design choice made by Q#. I think it's unlikely to change.

Reply:
Okay I think I'm getting it. So to summarize the difference
- `let` declares a container holding a value that may outlive the container and the current scope.
- `mutate` declares a container holding a value where the value may outlive the container, and the contain may outlive the value (for example if it is assigned another value).
- `use` declares a container holding a value that will end (is there a better way to phrase this?) when the container ends, which will be when the current scope ends.
- `borrow` is the same as `use` except the internal state of the value may be different.

Does that all seem correct?

Reply:
Oh would we then disallow:
```
use q = Qubit();
{
   use w = q;
}
```
In this case the value in the container `w` would be invalidated after the end of the inner scope, which would invalidate the value of `q` before `q` goes out of scope.

Reply:
The way I think about it is:

* `let` gives a name to a value. The name always refers to the same value.
* `mutable` gives a name to a value. The name can refer to different values at different times.
* `use` / `borrow` gives a name to a value. When the name goes out of scope, something like `Release(q)` is called for all qubits `q` contained in the value.

> Oh would we then disallow:
>
> ```qsharp
> use q = Qubit();
> {
>    use w = q;
> }
> ```

Because qubit initializer expressions can only contain `Qubit()`, `Qubit[n]`, and tuples of those, this is not allowed currently and is not allowed by this proposal, either.

Reply:
> It's pretty much the same as F# or C# -- something you "create" with `use` is `IDisposable` and gets disposed of (the `Dispose` method gets called) when it goes out of scope. We don't have interfaces, but qubits are the Q# equivalent of disposable.
> In all 3 languages, the `use` keyword on the left indicates that there's special processing of the _value_ (not the variable!) related to when the variable goes out of scope.

Two points by way of agreement:

- Python has very similar in the form of a `with` block:
  ```python
  with open('a.txt', 'r') as f:
      contents = f.read()
  # f and contents are both valid here, but the with block
  # ensures that f.__exit__ has been called, even in the case
  # of exceptional flow control.
  ```

- C# has recently introduced `using var` as a new kind of declaration that acts very similarly to the new `use` proposed here:
  ```csharp
  {
      // ...
      using var z = new SomethingDisposable();
      // ...
  } // z.Dispose will get called here, even with exceptional flow control.
  ```

Reply:
It would be good to summarize briefly the points on this thread either under the related mechanism section, to which we should add considerations around initializers in general.